### PR TITLE
Make sure the timestamps of the checkpoints are the same when loading

### PR DESCRIPTION
**Pull request by eracah**
# What does this PR do?
In the edge case where one is using sharded checkpointing + autoresume and one rank's save checkpoint operation fails, then that rank will resume from an older checkpoint than the other ranks and this will be a silent error. 

This PR fixes that by checking to make sure that all timestamps loaded in match. If they don't it errors out

### Manual Test
```image: mosaicml/pytorch:latest
integrations:
- git_branch: release/v0.0.4
  git_repo: mosaicml/examples
  integration_type: git_repo
  pip_install: .[llm]
- git_branch: ckpt-timestamp
  git_repo: eracah/evan-composer
  integration_type: git_repo
  pip_install: .[all]
parameters:
  load_path: s3://mosaicml-internal-checkpoints-test/evan-test/checkpoints/gpt_125m/sharded-ws-8/125m-ws-8-save-FYeuN0/ep0-ba1-rank{rank}.pt
  max_duration: 4ba
```
logs:
```
Starting training...
******************************
Config:
enabled_algorithms/GradientClipping: true
node_name: a100-80sxm-h10-03
num_gpus_per_node: 8
num_nodes: 1
rank_zero_seed: 17

******************************
[batch=2/4]:
         ...
[batch=3/4]:
         ....
[batch=4/4]:
         ...
```

# What issue(s) does this change relate to?
fix CO-2017

### CODE REVIEW

File: composer/utils/checkpoint.py [hunk 1]
```diff
@@ -484,6 +484,20 @@ def _restore_checkpoint(
             exclude_algorithms=exclude_algorithms,
             algorithm_passes=algorithm_passes,
         )
+        step_to_resume_from = state.timestamp.batch.value
+        avg_step_to_resume_from = state.device.tensor_to_device(
```
Comment (dakinggg) [inline composer/utils/checkpoint.py:488]: Any particular reason we're doing this with floats? the timestamp should be an int and then we don't have any risk of floating point error messing things up

Comment (eracah) [reply composer/utils/checkpoint.py:488]: oh I mainly did it to avoid rounding error when averaging.

Comment (eracah) [reply composer/utils/checkpoint.py]: ok changed to min-max comparison

**Review by dakinggg — APPROVED** [review]

**Comment by dakinggg** [comment]
Could you do a manual test too please? Preferable multi node, since we don't have unit test coverage for that. Would like to see a normal resumption succeed.

**Comment by eracah** [comment]
> Could you do a manual test too please? Preferable multi node, since we don't have unit test coverage for that. Would like to see a normal resumption succeed.

Ok did a standard resumption of a sharded checkpoint and added the setup and logs to the PR description

**Pull request merged**